src/script_lua.c

```diff
@@ -1260,13 +1283,16 @@ void luaCallFunction(scriptRunCtx* run_ctx, lua_State *lua, robj** keys, size_t
 
     /* Populate the argv and keys table accordingly to the arguments that
      * EVAL received. */
-    luaSetGlobalArray(lua,"KEYS",keys,nkeys);
-    luaSetGlobalArray(lua,"ARGV",args,nargs);
+    luaCreateArray(lua,keys,nkeys);
+    /* On eval, keys and arguments are globals. */
+    if (run_ctx->flags & SCRIPT_EVAL_MODE) lua_setglobal(lua,"KEYS");
+    luaCreateArray(lua,args,nargs);
+    if (run_ctx->flags & SCRIPT_EVAL_MODE) lua_setglobal(lua,"ARGV");
 
     /* At this point whether this script was never seen before or if it was
      * already defined, we can call it. We have zero arguments and expect
      * a single return value. */
-    int err = lua_pcall(lua,0,1,-2);
+    int err = lua_pcall(lua,run_ctx->flags & SCRIPT_EVAL_MODE ? 0 : 2,1,run_ctx->flags & SCRIPT_EVAL_MODE ? -2 : -4);
```

Comment:
maybe it would be more readable to have two calls for `lua_pcall`, each with a different set of arguments.

Reply:
Fixed